src/zope/browserpage/metaconfigure.py

```diff
@@ -411,7 +412,12 @@ def _handle_for(_context, for_):
 class simple(BrowserView):
 
     def publishTraverse(self, request, name):
-        raise NotFound(self, name, request)
+        if name in getattr(self, "_simple_whitelist", []):
+            self.__page_attribute__ = name
+            self.__call__ = simple.__call__.__get__(self, simple)
```

Comment:
`simple` is added as a base alongside whatever class the user defined. If the user defined `__call__` then this will win, over the one in `simple`. We need the `__call__` attribute to point to the way of rendering the view, so the version in `simple` actually is perfect for what we need. In this case, we want to make the instance use the `simple.__call__` implementation rather than the one the class alone would pick.

The way I do this is by setting `__call__` on the instance to point at the right function. In Python3 we can just do `self.__call__ = simple.__call__` because in Python3 `type(simple.__call__)` is `<class 'function'>`. In Python2 it's `<type 'instancemethod'>`. This is to do with the whole new/old style class compatibility thing back in Python2.

This line uses the fact that instancemethods and functions are descriptors to explicitly bind it to the instance when setting it on the instance dictionary. If we don't do this then `self` looks like just another parameter to Python2 as it's not gone through the normal method resolution protocol.

Reply:
Would `self.__call__ = self.__call__`  work to invoke the descriptor protocol?

But I guess my main question was about why setting `__call__` on an instance even works?  Is the publisher doing a getattr(obj, '__call__') explicitly?  Because Python itself won't use `__call__` of an instance, it'll only look up special methods from the class.

Reply:
BTW all the tests pass if I comment the `self.__call__` assignment, so either it's not necessary, or the tests are inadequate.

Reply:
Maybe `self.__call__ = self.__call__` would work, I'm not sure. It looks even weirder to me, though.

Yes, it works because there's an explicit getattr for call. This is actually done in code that's relevant to this PR, in https://github.com/zopefoundation/zope.browserpage/blob/master/src/zope/browserpage/metaconfigure.py#L146 so we could make this a lower impact change by changing the relevant method to a dispatcher and allowing it to call the discovered `__call__`.

Yeah, I've only added pretty minimal tests. To be honest, this is a really busy time for me and I jumped in on this because I was directly emailed a couple of times to ask for my opinion. I meant this more as a demonstration of the approach I would take than a production-ready PR. I'm happy to update it but equally I'm happy for people to just change the branches or branch themselves to get something they're confident in.

Reply:
Could it be that this assignment is only necessary if a child class has its own `__call__` to reset it to actually get the attribute? Maybe there should be a test demonstrating its necessity.